### PR TITLE
Fix MSVC build after major change

### PR DESCRIPTION
**Pull request by peterjc123**
1. Add export macros definition for the export modifier to work
2. Fix some functions or constants that differs from Unix
3. Add some include files and build scipts

### CODE REVIEW

**Comment by ezyang** [comment]
Thanks @peterjc123, you are a trooper!

File: aten/src/ATen/Half.h [hunk 1]
```diff
@@ -10,6 +10,15 @@
 #include <cuda_fp16.h>
 #endif
 
+#ifdef _MSC_VER
+// MSVC does not support std::inf and std::nan for integral types
+#define ISINF(x) std::isinf((double)x)
+#define ISNAN(x) std::isnan((double)x)
```
Comment (apaszke) [inline aten/src/ATen/Half.h]: I think a better fix would be to always return `false` for integral types (should be possible to do inside this macro with some template magic). Not all values of `long long` can be represented in `double`

**Review by apaszke — COMMENTED** [review]
Looks good, but please change `ISINF` and `ISNAN` to skip the check and return `false` for integral types

**Comment by peterjc123** [comment]
@apaszke Updated using `enable_if` magics.

File: aten/src/ATen/Half.h [hunk 2]
```diff
@@ -16,15 +16,23 @@ template<typename To, typename From> To convert(From f) {
...
+template<typename To, typename From>
+typename std::enable_if<!std::is_integral<From>::value, bool>::type overflows(From f) {
+	using limit = std::numeric_limits<To>;
+	if (limit::has_infinity && std::isinf(f)) {
+		return false;
+	}
+	if (!limit::has_quiet_NaN && std::isnan(f)) {
+		return true;
+	}
+	return f < limit::lowest() || f > limit::max();
```
Comment (apaszke) [inline aten/src/ATen/Half.h]: You're using tabs. Please change it to spaces

**Review by apaszke — COMMENTED** [review]
Looks good for me. cc @zdevito for approving ATen changes

**Comment by peterjc123** [comment]
Rebased.

File: aten/CMakeLists.txt [hunk 3]
```diff
@@ -43,7 +43,6 @@ IF(MSVC)
   # we want to respect the standard, and we are bored of those **** .
   ADD_DEFINITIONS(-D_CRT_SECURE_NO_DEPRECATE=1)
   LIST(APPEND CUDA_NVCC_FLAGS "-Xcompiler /wd4819 -Xcompiler /wd4503")
-  ADD_DEFINITIONS(-DTH_EXPORTS)
```
Comment (zdevito) [inline aten/CMakeLists.txt]: Is there a reason this has to be in specific subfolders now? I'd expect it to only change the build command if it were set here.

**Comment by peterjc123** [comment]
Now it should be ok.

**Comment by soumith** [comment]
thanks @peterjc123 !

**Pull request merged**